samza-core/src/main/java/org/apache/samza/clustermanager/ResourceRequestState.java

```diff
@@ -213,7 +226,7 @@ public void updateStateAfterAssignment(SamzaResourceRequest request, String assi
   public int releaseExtraResources() {
     synchronized (lock) {
       int numReleasedResources = 0;
-      if (requestsQueue.isEmpty()) {
+      if (requestsQueue.isEmpty() && delayedRequestsQueue.isEmpty()) {
```

Comment:
We don't need to check the delayedRequestsQueue here, because a resource-request for anything in the delayedRequestsQueue will be sent out only when the requests are no-longer delayed.

The implication of this is that if requestsQueue is empty and delayedRequestsQueue is not, the CPM will continue to hold onto allocated resources, in the worst case for a period of 5 mins (default). 
Aggregated over num_containers, could be significant especially in resource-crunch scenarios.

Alternatively, we could release the resource only looking at the requestsQueue, and when a request in the delayedRequestsQueue "expires", we will send out the request to the YARN-RM and allocation flow shall resume.